utils/serde/src/main/java/io/apicurio/registry/utils/serde/AbstractKafkaDeserializer.java

```diff
@@ -76,4 +82,40 @@ public U deserialize(String topic, byte[] data) {
         int start = buffer.position() + buffer.arrayOffset();
         return readData(schema, buffer, start, length);
     }
+
+    @Override
+    public U deserialize(String topic, Headers headers, byte[] data) {
+        if (data == null) {
+            return null;
+        }
+        // check if data contains the magic byte
+        if (data[0] == MAGIC_BYTE){
+            return deserialize(topic, data);
+        } else {
+            Long id = headerUtils.getGlobalId(headers);
+            if (id == null) {
+                String artifactId = headerUtils.getArtifactId(headers);
+                Integer version = headerUtils.getVersion(headers);
+                id = toGlobalId(artifactId, version);
+            }
+            T schema = getCache().getSchema(id);
+            ByteBuffer buffer = ByteBuffer.wrap(data);
+            int length = buffer.limit();
+            int start = buffer.position();
+            return readData(headers, schema, buffer, start, length);
+        }
+    }
+
+    protected Long toGlobalId(String artifactId, Integer version) {
+        if (artifactId == null) {
+            throw new RuntimeException("ArtifactId not found in headers.");
+        }
+        if (version == null) {
+            ArtifactMetaData amd = getClient().getArtifactMetaData(artifactId);
+            return amd.getGlobalId();
+        } else {
+            VersionMetaData vmd = getClient().getArtifactVersionMetaData(version, artifactId);
+            return vmd.getGlobalId();
+        }
+    }
```

Comment:
Some of this logic is duplicated in `JsonSchemaKafkaDeserializer` I think - is there an opportunity to consolidate this?